admin/database/postgres/postgres.go

```diff
@@ -132,3 +133,95 @@ func (c *connection) DeleteProject(ctx context.Context, id string) error {
 	_, err := c.db.ExecContext(ctx, "DELETE FROM projects WHERE id=$1", id)
 	return err
 }
+
+// InsertAuthCode inserts the authorization code data into the store.
+func (c *connection) InsertAuthCode(ctx context.Context, code *database.AuthCode) error {
+	res, err := c.db.ExecContext(ctx,
+		"INSERT INTO device_code_auth (device_code, user_code, expires_at, approval_state, client_id, user_id, used) VALUES ($1, $2, $3, $4, $5, $6, $7)",
+		code.DeviceCode, code.UserCode, code.Expiry, code.ApprovalState, code.ClientID, code.UserID, code.Used)
+	if err != nil {
+		return err
+	}
+	rows, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rows != 1 {
+		return fmt.Errorf("could not insert auth code, expected 1 row to be affected, got %d", rows)
+	}
+	return nil
+}
+
+// GetByDeviceCode retrieves the authorization code data from the store
+func (c *connection) GetByDeviceCode(ctx context.Context, deviceCode string) (*database.AuthCode, error) {
+	authCode := &database.AuthCode{}
+	err := c.db.QueryRowxContext(ctx, "SELECT * FROM device_code_auth WHERE device_code = $1", deviceCode).StructScan(authCode)
+	if err != nil {
+		return nil, err
+	}
+	return authCode, nil
+}
+
+// GetByUserCode retrieves the authorization code data from the store
+func (c *connection) GetByUserCode(ctx context.Context, userCode string) (*database.AuthCode, error) {
+	authCode := &database.AuthCode{}
+	err := c.db.QueryRowxContext(ctx, "SELECT * FROM device_code_auth WHERE user_code = $1", userCode).StructScan(authCode)
+	if err != nil {
+		return nil, err
+	}
+	return authCode, nil
```

Comment:
shouldn't  this `.StructScan(authCode)` throw err when no rows found ? Anyways will add a check

Reply:
actually its checked in the caller to provide meaningful err to the caller and not send internal server err. If there is another we send internal server err, see here - https://github.com/rilldata/rill-developer/pull/1858/files#diff-0232f89fb0b53279484876cc4f7e29ec2ca9205be0a24dde55bb6838734adb36R142

Reply:
Yeah it will, I just meant to map `sql.ErrNoRows` to `database.ErrNotFound`, like here: https://github.com/rilldata/rill-developer/blob/40f19898a7d7550b7ec2689fc5fdfd6a6fa8cc91/admin/database/postgres/postgres.go#L62

Maybe the `database.ErrNotFound` type is a bit abstraction overkill, but then we should deprecate it entirely and always return `sql.ErrNoRows`. Probably should be a separate PR then